src/switchroot/ostree-prepare-root.c

```diff
@@ -385,54 +413,67 @@ main (int argc, char *argv[])
         }
-
-      if (composefs_mode == OSTREE_COMPOSEFS_MODE_MAYBE)
-        g_print ("Trying to mount composefs rootfs\n");
-      else if (composefs_digest != NULL)
-        g_print ("Mounting composefs rootfs with expected digest '%s'\n", composefs_digest);
       else
-        g_print ("Mounting composefs rootfs\n");
+        {
+          // If we're not verifying a digest, then we *must* also have signatures disabled.
+          // Or stated in reverse: if signature verification is enabled, then digest verification
+          // must also be.
+          g_assert (!composefs_config->signature_pubkey);
+          g_print ("composefs: Mounting with no digest or signature check\n");
+        }
 
       if (lcfs_mount_image (OSTREE_COMPOSEFS_NAME, TMP_SYSROOT, &cfs_options) == 0)
         {
-          using_composefs = 1;
+          using_composefs = true;
           g_variant_builder_add (&metadata_builder, "{sv}", OTCORE_RUN_BOOTED_KEY_COMPOSEFS,
                                  g_variant_new_boolean (true));
         }
       else
         {
-          if (errno == ENOVERITY)
-            g_print ("No verity in composefs image\n");
-          else if (errno == EWRONGVERITY)
-            g_print ("Wrong verity digest in composefs image\n");
-          else if (errno == ENOSIGNATURE)
-            g_print ("Missing signature in composefs image\n");
+          int errsv = errno;
+          const char *errmsg;
+          switch (errsv)
+            {
+            case ENOVERITY:
+              errmsg = "fsverity not enabled on composefs image";
+              break;
+            case EWRONGVERITY:
+              errmsg = "Wrong fsverity digest in composefs image";
+              break;
+            case ENOSIGNATURE:
+              errmsg = "Missing signature for fsverity in composefs image";
+              break;
+            default:
+              errmsg = strerror (errno);
+              break;
+            }
+          if (composefs_config->enabled == OT_TRISTATE_MAYBE)
+            {
+              g_print ("composefs: optional support failed: %s\n", errmsg);
+            }
           else
-            g_print ("Mounting composefs image failed: %s\n", strerror (errno));
+            {
+              g_assert (composefs_config->enabled == OT_TRISTATE_YES);
+              errx (EXIT_FAILURE, "composefs: failed to mount: %s", errmsg);
+            }
         }
 #else
-      err (EXIT_FAILURE, "Composefs not supported");
+      errx (EXIT_FAILURE, "composefs: enabled at runtime, but support is not compiled in");
 #endif
     }
 
   if (!using_composefs)
     {
-      if (composefs_mode > OSTREE_COMPOSEFS_MODE_MAYBE)
-        err (EXIT_FAILURE, "Failed to mount composefs");
-
       /* The deploy root starts out bind mounted to sysroot.tmp */
       if (mount (deploy_path, TMP_SYSROOT, NULL, MS_BIND | MS_SILENT, NULL) < 0)
         err (EXIT_FAILURE, "failed to make initial bind mount %s", deploy_path);
     }
-  else
```

Comment:
I wonder should we have log levels in prepare-root, "Mounted composefs" would be a nice example of a verbose or debug thing

Reply:
Yes, fair point.  I did move the "mounted composefs" metadata into the `/run/ostree-booted` in the previous PR where it's introspectable in a more reliable fashion than log messages.  But this one is certainly worthy of a log message...I'll re-add.